File: src/activities/git_operations.py
"""Git operations activities for Temporal workflows."""

import logging
import os
from dataclasses import dataclass
from pathlib import Path
from typing import Optional

from git import Repo, GitCommandError
from temporalio import activity

logger = logging.getLogger(__name__)


@dataclass
class CloneResult:
    """Result of a repository clone operation."""
    workspace_path: str
    success: bool
    error: Optional[str] = None


@dataclass
class BranchResult:
    """Result of a branch creation operation."""
    branch_name: str
    success: bool
    error: Optional[str] = None


@dataclass
class CommitResult:
    """Result of a commit operation."""
    commit_sha: str
    success: bool
    error: Optional[str] = None


@dataclass
class PushResult:
    """Result of a push operation."""
    success: bool
    error: Optional[str] = None


@activity.defn
async def clone_repository(repo_url: str, workspace_dir: str) -> CloneResult:
    """
    Clone a Git repository to a workspace directory.

    Args:
        repo_url: The URL of the repository to clone
        workspace_dir: The directory path where the repository will be cloned

    Returns:
        CloneResult with workspace path and operation status
    """
    try:
        logger.info(f"Cloning repository {repo_url} to {workspace_dir}")

        # Ensure the parent directory exists
        workspace_path = Path(workspace_dir)
        workspace_path.parent.mkdir(parents=True, exist_ok=True)

        # Clone the repository
        repo = Repo.clone_from(repo_url, workspace_dir)

        logger.info(f"Successfully cloned repository to {workspace_dir}")
        return CloneResult(
            workspace_path=str(workspace_path.absolute()),
            success=True
        )

    except GitCommandError as e:
        error_msg = f"Git command failed: {e}"
        logger.error(error_msg)
        return CloneResult(
            workspace_path="",
            success=False,
            error=error_msg
        )
    except Exception as e:
        error_msg = f"Failed to clone repository: {e}"
        logger.error(error_msg)
        return CloneResult(
            workspace_path="",
            success=False,
            error=error_msg
        )


@activity.defn
async def create_branch(workspace: str, branch_name: str) -> BranchResult:
    """
    Create a new Git branch in the workspace.

    Args:
        workspace: The path to the Git repository workspace
        branch_name: The name of the branch to create

    Returns:
        BranchResult with branch name and operation status
    """
    try:
        logger.info(f"Creating branch {branch_name} in {workspace}")

        # Open the repository
        repo = Repo(workspace)

        # Create and checkout the new branch
        new_branch = repo.create_head(branch_name)
        new_branch.checkout()

        logger.info(f"Successfully created and checked out branch {branch_name}")
        return BranchResult(
            branch_name=branch_name,
            success=True
        )

    except GitCommandError as e:
        error_msg = f"Git command failed: {e}"
        logger.error(error_msg)
        return BranchResult(
            branch_name=branch_name,
            success=False,
            error=error_msg
        )
    except Exception as e:
        error_msg = f"Failed to create branch: {e}"
        logger.error(error_msg)
        return BranchResult(
            branch_name=branch_name,
            success=False,
            error=error_msg
        )


@activity.defn
async def commit_changes(workspace: str, message: str) -> CommitResult:
    """
    Commit all changes in the workspace.

    Args:
        workspace: The path to the Git repository workspace
        message: The commit message

    Returns:
        CommitResult with commit SHA and operation status
    """
    try:
        logger.info(f"Committing changes in {workspace}")

        # Open the repository
        repo = Repo(workspace)

        # Stage all changes
        repo.git.add(A=True)

        # Check if there are changes to commit
        if not repo.index.diff("HEAD") and not repo.untracked_files:
            logger.info("No changes to commit")
            return CommitResult(
                commit_sha="",
                success=True,
                error="No changes to commit"
            )

        # Commit the changes
        commit = repo.index.commit(message)

        logger.info(f"Successfully committed changes with SHA {commit.hexsha}")
        return CommitResult(
            commit_sha=commit.hexsha,
            success=True
        )

    except GitCommandError as e:
        error_msg = f"Git command failed: {e}"
        logger.error(error_msg)
        return CommitResult(
            commit_sha="",
            success=False,
            error=error_msg
        )
    except Exception as e:
        error_msg = f"Failed to commit changes: {e}"
        logger.error(error_msg)
        return CommitResult(
            commit_sha="",
            success=False,
            error=error_msg
        )


@activity.defn
async def push_changes(workspace: str) -> PushResult:
    """
    Push committed changes to the remote repository.

    Args:
        workspace: The path to the Git repository workspace

    Returns:
        PushResult with operation status
    """
    try:
        logger.info(f"Pushing changes from {workspace}")

        # Open the repository
        repo = Repo(workspace)

        # Get the current branch
        current_branch = repo.active_branch

        # Push to origin
        origin = repo.remote(name="origin")
        origin.push(current_branch.name)

        logger.info(f"Successfully pushed branch {current_branch.name}")
        return PushResult(success=True)

    except GitCommandError as e:
        error_msg = f"Git command failed: {e}"
        logger.error(error_msg)
        return PushResult(
            success=False,
            error=error_msg
        )
    except Exception as e:
        error_msg = f"Failed to push changes: {e}"
        logger.error(error_msg)
        return PushResult(
            success=False,
            error=error_msg
        )

File: internal/activities/git_operations.go
package activities

import (
	"context"
	"fmt"
	"log/slog"
	"os"
	"path/filepath"

	"github.com/go-git/go-git/v5"
	"github.com/go-git/go-git/v5/config"
	"github.com/go-git/go-git/v5/plumbing"
	"github.com/go-git/go-git/v5/plumbing/transport/http"
)

type GitActivities struct {
	GitHubToken string
}

type CloneResult struct {
	WorkspacePath string `json:"workspace_path"`
	Success       bool   `json:"success"`
	Error         string `json:"error,omitempty"`
}

type BranchResult struct {
	BranchName string `json:"branch_name"`
	Success    bool   `json:"success"`
	Error      string `json:"error,omitempty"`
}

type CommitResult struct {
	CommitSHA string `json:"commit_sha"`
	Success   bool   `json:"success"`
	Error     string `json:"error,omitempty"`
}

type PushResult struct {
	Success bool   `json:"success"`
	Error   string `json:"error,omitempty"`
}

func (a *GitActivities) CloneRepository(ctx context.Context, repoURL, workspaceDir string) (CloneResult, error) {
	slog.Info("Cloning repository", "url", repoURL, "dir", workspaceDir)

	parentDir := filepath.Dir(workspaceDir)
	if err := os.MkdirAll(parentDir, 0o755); err != nil {
		return CloneResult{}, fmt.Errorf("creating parent directory: %w", err)
	}

	cloneOpts := &git.CloneOptions{
		URL: repoURL,
	}
	if a.GitHubToken != "" {
		cloneOpts.Auth = &http.BasicAuth{
			Username: "x-access-token",
			Password: a.GitHubToken,
		}
	}

	_, err := git.PlainCloneContext(ctx, workspaceDir, false, cloneOpts)
	if err != nil {
		return CloneResult{}, fmt.Errorf("cloning repository: %w", err)
	}

	absPath, _ := filepath.Abs(workspaceDir)
	slog.Info("Repository cloned", "path", absPath)

	return CloneResult{
		WorkspacePath: absPath,
		Success:       true,
	}, nil
}

func (a *GitActivities) CreateBranch(ctx context.Context, workspace, branchName string) (BranchResult, error) {
	slog.Info("Creating branch", "branch", branchName, "workspace", workspace)

	repo, err := git.PlainOpen(workspace)
	if err != nil {
		return BranchResult{}, fmt.Errorf("opening repository: %w", err)
	}

	wt, err := repo.Worktree()
	if err != nil {
		return BranchResult{}, fmt.Errorf("getting worktree: %w", err)
	}

	err = wt.Checkout(&git.CheckoutOptions{
		Branch: plumbing.NewBranchReferenceName(branchName),
		Create: true,
	})
	if err != nil {
		return BranchResult{}, fmt.Errorf("creating branch: %w", err)
	}

	slog.Info("Branch created", "branch", branchName)
	return BranchResult{
		BranchName: branchName,
		Success:    true,
	}, nil
}

func (a *GitActivities) CommitChanges(ctx context.Context, workspace, message string) (CommitResult, error) {
	slog.Info("Committing changes", "workspace", workspace)

	repo, err := git.PlainOpen(workspace)
	if err != nil {
		return CommitResult{}, fmt.Errorf("opening repository: %w", err)
	}

	wt, err := repo.Worktree()
	if err != nil {
		return CommitResult{}, fmt.Errorf("getting worktree: %w", err)
	}

	_, err = wt.Add(".")
	if err != nil {
		return CommitResult{}, fmt.Errorf("staging changes: %w", err)
	}

	status, err := wt.Status()
	if err != nil {
		return CommitResult{}, fmt.Errorf("getting status: %w", err)
	}

	if status.IsClean() {
		slog.Info("No changes to commit")
		return CommitResult{
			Success: true,
			Error:   "No changes to commit",
		}, nil
	}

	hash, err := wt.Commit(message, &git.CommitOptions{})
	if err != nil {
		return CommitResult{}, fmt.Errorf("committing changes: %w", err)
	}

	slog.Info("Changes committed", "sha", hash.String())
	return CommitResult{
		CommitSHA: hash.String(),
		Success:   true,
	}, nil
}

func (a *GitActivities) PushChanges(ctx context.Context, workspace string) (PushResult, error) {
	slog.Info("Pushing changes", "workspace", workspace)

	repo, err := git.PlainOpen(workspace)
	if err != nil {
		return PushResult{}, fmt.Errorf("opening repository: %w", err)
	}

	head, err := repo.Head()
	if err != nil {
		return PushResult{}, fmt.Errorf("getting HEAD: %w", err)
	}

	pushOpts := &git.PushOptions{
		RemoteName: "origin",
		RefSpecs:   []config.RefSpec{config.RefSpec(head.Name() + ":" + head.Name())},
	}
	if a.GitHubToken != "" {
		pushOpts.Auth = &http.BasicAuth{
			Username: "x-access-token",
			Password: a.GitHubToken,
		}
	}

	err = repo.PushContext(ctx, pushOpts)
	if err != nil {
		return PushResult{}, fmt.Errorf("pushing changes: %w", err)
	}

	slog.Info("Changes pushed")
	return PushResult{Success: true}, nil
}
